Dev container (#44)
nvidia-holoscan/holohub: README.md

HoloHub is based on [Holoscan SDK](https://github.com/nvidia-holoscan/holoscan-sdk).
HoloHub has been tested and is known to run on Ubuntu 20.04. Other versions of Ubuntu or OS distributions may result in build and/or runtime issues.

1. Clone this repository.

2. Choose to build Holohub using development container or using a local environment.

> **Container build**

Reviewer: ```suggestion
### Container build
```

> ***Build container:*** Run the following command from the holohub directory to build the development container.

Reviewer: ```suggestion
***Building dev container:*** Run the following command from the holohub directory to build the development container.
```

```bash
  ./dev_container build
```

***Note:*** The development container script ```dev_container``` will by default use [NGC's Holoscan SDK container](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/containers/holoscan) for the local GPU configuration by detecting if the system is using an iGPU (integrated GPU) or a dGPU (discrete GPU). 

> For example on x86 ```docker images``` will list the following new images after ```./dev_container build``` has completed.

Reviewer: ```suggestion
For example, on a system with dGPU ```docker images``` will list the following new images after ```./dev_container build``` has completed.
```

```bash
user@ubuntu-20-04:/media/data/github/holohub$ docker images
REPOSITORY                               TAG           IMAGE ID       CREATED         SIZE
holohub                                  ngc-v0.5.1-dgpu   b6d86bccdcac   9 seconds ago   8.47GB
nvcr.io/nvidia/clara-holoscan/holoscan   v0.5.1-dgpu       1b4df7733d5b   5 weeks ago     8.04GB
```

***Advanced build options***

1. Custom base image

> It is possible to configure a custom base image for building the Holohub container. E.g., if you built Holoscan SDK locally and want to use the locally built container as base image, use the following command: 

Reviewer: ```suggestion
It is possible to configure a custom base image for building the Holohub container. E.g., if you built Holoscan SDK locally and want to use the locally built container as the base image, use the following command: 
```

```bash

>   ./dev_container build --base_img holoscan-sdk-dev:latest --img_fqn holohub:sdk-dev-latest

Reviewer: ```suggestion
  ./dev_container build --base_img holoscan-sdk-dev:latest --img holohub:sdk-dev-latest
```
Probably don't need to add 'fqn' to the option name

```

> where ```--base_img```  is used to configure the base container image and ```--img_fqn``` to define the fully qualified image name. 

Reviewer: ```suggestion
where ```--base_img```  is used to configure the base container image and ```--img``` to define the fully qualified image name. 
```
Probably don't need 'fqn' in the option name

After ```./dev_container build``` has completed ```docker images``` will list the following new image

```bash
user@ubuntu-20-04:/media/data/github/holohub$ docker images
REPOSITORY                               TAG               IMAGE ID       CREATED          SIZE
holohub                                  sdk-dev-latest    cb0231f77856   54 seconds ago   8.22GB
```

2. Using custom Dockerfile

```bash

>   ./dev_container build  --docker_file PATH_TO_DOCKERFILE  --img_fqn holohub-debug:local-sdk-v0.5.1

Reviewer: ```suggestion
  ./dev_container build  --docker_file <path_to_dockerfile>  --img holohub-debug:local-sdk-v0.5.1
```
Use similar style as other parts of holohub docs

```

> where ```--docker_file```  is the path to the container's Dockerfile and ```--img_fqn``` to define the fully qualified image name. 

Reviewer: ```suggestion
where ```--docker_file```  is the path to the container's Dockerfile and ```--img``` to define the fully qualified image name. 
```

***Note***:  To debug the values for base image, docker file, gpu type and output image name use ```--debug```. 

For example, on x86 with dGPU the default build command will print the following values when using the ```--debug``` option. 

```bash
user@ubuntu-20-04:/media/data/github/holohub$ ./dev_container build --debug
Build (HOLOHUB_ROOT:/media/data/github/holohub)...
Build (gpu_type_type:dgpu)...
Build (base_img:nvcr.io/nvidia/clara-holoscan/holoscan:v0.5.1-dgpu)...
Build (docker_file_path:/media/data/github/holohub/Dockerfile)...
Build (img_fqn:holohub:ngc-v0.5.1-dgpu)...
....
```

> ***Launch container:***  Run the following command from the holohub directory to launch the default development container built using Holoscan SDK's container from ngc for the local GPU.

Reviewer: ```suggestion
***Launching dev container:***  Run the following command from the holohub directory to launch the default development container built using Holoscan SDK's container from ngc for the local GPU.
```

```bash
  ./dev_container launch
```

> The launch script will also inspect for available video devices (V4L2, AJA capture boards, Deltacast capture boards) and presence of Deltacast's Videomaster SDK and map it into the development container.

Reviewer: ```suggestion
The launch script will also inspect for available video devices (V4L2, AJA capture boards, Deltacast capture boards) and the presence of Deltacast's Videomaster SDK and map it into the development container.
```

***Advanced launch options***

1. For Holohub image built with locally built Holoscan SDK container

```bash

>   ./dev_container launch --img holohub:local-sdk-latest --local_sdk_root PATH_TO_HOLOSCAN_SDK

Reviewer: ```suggestion
  ./dev_container launch --img holohub:local-sdk-latest --local_sdk_root <path_to_holoscan_sdk>
```

```

2. Launch custom Holohub container with fully qualified name, e.g. "holohub:ngc-sdk-sample-app"

```bash
  ./dev_container launch --img holohub:ngc-sdk-sample-app
```

> ***Note***:  To debug the values passed to the docker run command when using add the ```--debug``` option to the launch command.

Reviewer: ```suggestion
***Note***:  To print the values passed to the docker run command, add the ```--debug``` option to the launch command.
```

> For example, on x86 with dGPU ```./dev_container launch --debug``` will print the following values. 

Reviewer: ```suggestion
For example, on an x86_64 system with dGPU ```./dev_container launch --debug``` will print the following values. 
```

Launch (local_sdk_opt: )...
Launch (nvidia_icd_json: /usr/share/vulkan/icd.d/nvidia_icd.json)...
Launch (image: holohub:ngc-v0.5.1-dgpu)...
....
```

Please note that the values of some of the variables will vary depending on configured options, iGPU or dGPU, availability of devices for video capture, ... 

From within the container build the Holohub apps as explained in section [Building HoloHub](#building-holohub).

The development container has been tested on the following platforms: x86_x64 workstation with multiple RTX GPUs, Clara AGX Dev Kit (dGPU mode), IGX Orin Dev Kit (dGPU mode), AGX Orin Dev Kit (iGPU).

***Notes for AGX Orin Dev Kit***: 

(1) On AGX Orin Dev Kit the launch script will add ```--privileged``` and ```--group-add video``` to the docker run command for the Holohub sample apps to work. Please also make sure that the current user is member of the group video.  

(2) When building Holoscan SDK on AGX Orin Dev Kit from source please add the option  ```--cudaarchs all``` to the ```./run build``` command to include support for AGX Orin's iGPU.

> **Local build** 

Reviewer: ```suggestion
### Local build
```
